code/datums/ai/ai_controller.dm

```diff
@@ -0,0 +1,753 @@
+	. = AI_STATUS_ON
+
+	if(!ismob(pawn))
+		return
+
+	var/mob/living/mob_pawn = pawn
+	if(!continue_processing_when_client && mob_pawn.client)
+		. = AI_STATUS_OFF
+
+	if(ai_traits & CAN_ACT_WHILE_DEAD)
+		return
+
+	if(mob_pawn.stat == DEAD)
+		. = AI_STATUS_OFF
+
+	var/turf/pawn_turf = get_turf(mob_pawn)
+#ifdef GAME_TESTS
+	if(!pawn_turf)
+		CRASH("AI controller [src] controlling pawn ([pawn]) is not on a turf.")
+#endif
+	if(!length(SSmobs.clients_by_zlevel[pawn_turf.z]))
+		. = AI_STATUS_OFF
+
+/// Called when the AI controller pawn changes z levels.
+/// We check if there's any clients on the new one and wake up the AI if there is.
+/datum/ai_controller/proc/on_changed_z_level(atom/source, turf/old_turf, turf/new_turf, same_z_layer, notify_contents)
+	SIGNAL_HANDLER // COMSIG_MOVABLE_Z_CHANGED
+
+	if(ismob(pawn))
+		var/mob/mob_pawn = pawn
+		if(mob_pawn?.client && !continue_processing_when_client)
+			return
+	if(old_turf)
+		SSai_controllers.ai_controllers_by_zlevel[old_turf.z] -= src
+	if(new_turf)
+		SSai_controllers.ai_controllers_by_zlevel[new_turf.z] += src
+		var/new_level_clients = length(SSmobs.clients_by_zlevel[new_turf.z])
+		if(new_level_clients)
+			set_ai_status(AI_STATUS_IDLE)
+		else
+			set_ai_status(AI_STATUS_OFF)
+
+/// Abstract proc for initializing the pawn to the new controller
+/datum/ai_controller/proc/try_possess_pawn(atom/new_pawn)
+	return
+
+/// Proc for deinitializing the pawn to the old controller
+/datum/ai_controller/proc/unpossess_pawn(destroy)
+	if(isnull(pawn))
+		return // instantiated without an applicable pawn, fine
+
+	UnregisterSignal(pawn, list(COMSIG_MOB_LOGIN, COMSIG_MOB_LOGOUT, COMSIG_MOB_STATCHANGE, COMSIG_PARENT_QDELETING))
+	if(ai_movement.moving_controllers[src])
+		ai_movement.stop_moving_towards(src)
+	var/turf/pawn_turf = get_turf(pawn)
+	if(pawn_turf)
+		SSai_controllers.ai_controllers_by_zlevel[pawn_turf.z] -= src
+	if(ai_status)
+		SSai_controllers.ai_controllers_by_status[ai_status] -= src
+	pawn.ai_controller = null
+	pawn = null
+	if(destroy)
+		qdel(src)
+
+/// Returns TRUE if the ai controller can actually run at the moment.
+/datum/ai_controller/proc/able_to_run()
+	if(HAS_TRAIT(pawn, TRAIT_AI_PAUSED))
+		return FALSE
+	if(world.time < paused_until)
+		return FALSE
+	return TRUE
+
+
+/// Runs any actions that are currently running
+/datum/ai_controller/process(seconds_per_tick)
+
+	if(!able_to_run())
+		GLOB.move_manager.stop_looping(pawn) //stop moving
+		return //this should remove them from processing in the future through event-based stuff.
+
+	if(!LAZYLEN(current_behaviors) && idle_behavior)
+		idle_behavior.perform_idle_behavior(seconds_per_tick, src) //Do some stupid shit while we have nothing to do
+		return
+
+	if(current_movement_target)
+		if(!isatom(current_movement_target))
+			stack_trace("[pawn]'s current movement target is not an atom, rather a [current_movement_target.type]! Did you accidentally set it to a weakref?")
```

Comment:
We don't really have/use weakrefs like /tg/ does here; since we use UIDs, I think the `.type` call may also runtime. Perhaps just take its base string representation

Reply:
Yeah, the rest of the PR was converted from weakrefs to UIDs, I just forgot this piece.